### ISSUE TITLE
[User Story] Collection and classification of new kernel needs

### ISSUE THREAD

**Issue by learning-chip** [huawei-csl/pto-kernels#23]
We keep getting new kernels needs from customers. Here try to classify and keep track of them.

**Classify by new model architecture:**
- Sparse Attention, in particular the [DeepSeek Sparse Attention](https://gitcode.com/cann/cann-recipes-infer/tree/master/docs/models/deepseek-v3.2-exp), used by [DeepSeek-V3.2](https://huggingface.co/deepseek-ai/DeepSeek-V3.2), [GLM-5](https://huggingface.co/zai-org/GLM-5)
- Linear attention, in particular [GatedDeltaNet](https://arxiv.org/abs/2412.06464), used by [Qwen3-Coder-Next](https://huggingface.co/Qwen/Qwen3-Coder-Next), [Qwen3.5](https://huggingface.co/collections/Qwen/qwen35), [Kimi-Linear](https://huggingface.co/moonshotai/Kimi-Linear-48B-A3B-Instruct)
- "Classic" modules like fused grouped GEMM for MoE, [ref SonicMoE paper](https://arxiv.org/abs/2512.14080)

**Classify by scenario:**
- For inference: 
  - need dynamic batch dim and variable-length sequence dim
  - need paged KV cache, for standard and sparse attentions
  - need 8bit and [4bit](https://gitcode.com/cann/pto-isa/issues/115) quantized version
- For training: 
  - need both forward and backward pass, ref [Liger-Kernel](https://github.com/linkedin/Liger-Kernel) and the backward `_bwd` kernels [in FLA](https://github.com/fla-org/flash-linear-attention/)
  - usually need fused compute and communication, ref [communication ISA](https://gitcode.com/cann/pto-isa/pull/17), also ref CANN-baseline such as [mm_all_reduce](https://gitcode.com/Ascend/op-plugin/blob/7.3.0/docs/context/torch_npu-npu_mm_all_reduce_base.md) (also benefits inference, just training needs even more communication)

**Requirements:**
- Absolute performance measurement wrt roofline peak, plot FLOP and BW util rates
- Compare perf & precision to available baselines [in CANN/torch_npu](https://gitcode.com/Ascend/op-plugin/blob/7.3.0/docs/menu_Pytorch_API.md), [in sgl_kernel_npu](https://github.com/sgl-project/sgl-kernel-npu/tree/main/python/sgl_kernel_npu) (usually in triton), or in [tilelang-ascend](https://github.com/tile-ai/tilelang-ascend/tree/ascendc_pto/examples) 
- Dynamic vs static shapes: only allow specializing compile-time shape for `hidden_dim` and `num_head`, but `batch` and `sequence` must be run-time argument.
- Tensor layout: Attention kernel should prefer directly input and output `BSND` (`[batch, seq, num_head, hidden]`), to match FFN layer's output/input. Applies to both linear and sparse variants.
- E2E verification: inference kernels should be usable by [vllm-ascend](https://github.com/vllm-project/vllm-ascend) & [sglang](https://github.com/sgl-project/sglang), training kernels should be usable by [MindSpeed-LLM](https://gitcode.com/Ascend/MindSpeed-LLM) or at least by [torch-npu FSDP](https://gitcode.com/Ascend/pytorch/tree/master/test/distributed)
- If using AI coding (either fully to partially), document the prompt, context, skills, for future reuse.
- Prefer Python frontend to use auto-sync, fallback to C++ if have to manually plan synchronization

## Execution plan & order

**already planned & WIP**

For linear attention arch:
- [x] fast inverse for GDN
- [ ] e2e GDN with fast inverse
- [ ] GDN backward

For quantized inference:
- [x] fast rotation (hadamard, block-wise)
- [ ] fused with all ops nearby, ref https://github.com/huawei-csl/pto-kernels/issues/1

**Manpower needed...**
- [ ] Forward + backward in [Liger-Kernel](https://github.com/linkedin/Liger-Kernel), especially all the simple vector kernels.
- [ ] Some basic fused comm + comp demo, such as matmul+allreduce
- [ ] sequence parallel GDN training, ref [ZeCO](https://arxiv.org/abs/2507.01004)

## Remaining issue

- [ ] Need access to A5 hardware to use FP8 and C-V data-path... The PTO-ISA api stays largely the same.
